### PR TITLE
[JENKINS-42214] SandboxInterceptor must account for static members being accessed via objects instead of class references

### PR DESCRIPTION
**Pull request by dwnusbaum**
See [JENKINS-42214](https://issues.jenkins-ci.org/browse/JENKINS-42214). Fixing the issue going forward is easy enough, but I need to check the upgrade behavior for existing field signatures in their `scriptApproval.xml` that would be considered invalid after this change.

### CODE REVIEW

**Review by jglick — APPROVED** [review]

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java:555]: We could put this methods in `Whitelist` itself, but then we would need to come up with a different name.

In retrospect I think we should not have bothered with an API-level distinction between static and instance fields and methods in `Whitelist` and just made implementations check modifiers themselves.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java:576]: Similarly, we could put this in `StaticWhitelist`, but we would need to change the name. Another option could be to change the behavior of the current `rejectMethod` and deprecate `rejectStaticMethod`, but we would need to take a closer look at code like https://github.com/jenkinsci/permissive-script-security-plugin/blob/45af9a18a739fbb275093da5effcf1d2b7637a05/src/main/java/org/jenkinsci/plugins/permissivescriptsecurity/PermissiveWhitelist.java#L112 to understand whether that change could cause issues.

File: src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java:916]: This corrects a mistake in 85d16bd851b355458f6396dbc4f92c3e0eebcd72. Thankfully it did not mask any existing issues.

File: src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java:1659]: All of these cases previously threw instances of `RejectedAccessException` with bogus non-static signatures.

File: src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java:1670]: All of these cases previously failed because bogus non-static signatures were used when checking allowlists.

File: src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java:1656]:
```suggestion
    @Issue("JENKINS-42214")
    @Test
```

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java:422]: Intentionally changing this to delayed?

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java:555]: I guess so. A bit tricky to change now though, since you would need to use `Util.isOverridden` to accommodate the fact that there are both implementors and (apparently) callers of `Whitelist` in other plugins.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java
Comment (dwnusbaum) [reply src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java:422]: Ah yeah, I forgot to note that this was also an oversight in https://github.com/jenkinsci/script-security-plugin/commit/85d16bd851b355458f6396dbc4f92c3e0eebcd72. I think the effect is that the attempt to make static fields accessible via attribute syntax (`.@`) in that commit did not work correctly when modifying the attribute, only when getting it. Before that commit it did not work in either case, and after this commit it will work in both cases.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java
Comment (dwnusbaum) [reply src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java:555]: Yeah, I do not really think that it is worth making API changes at this point.

**Comment by dwnusbaum** [comment]
Rebuilding.

Also, I checked the behavior for cases where users have an invalid approved signature in their `scriptApproval.xml`. Previously, the behavior was wrong, but it was wrong both when checking the signature against allowlists and when reporting the rejection, so things worked fine as long as you had assertions disabled. If you approved some invalid signature before this update, then after this update you will need to approve the correct signature.

We could check `Signature.exists` during deserialization and try to migrate existing signatures that hit this case, but I'm not sure that the complexity is worth it for what is probably an uncommon case. I think it should be good enough to mention that you may need to approve some signatures after the update in the release notes.

**Comment by jglick** [comment]
> Rebuilding

BTW **Re-run** from **Checks** should work.

**Comment by jglick** [comment]
> I think it should be good enough to mention that you may need to approve some signatures after the update in the release notes.

Agreed

**Comment by dwnusbaum** [comment]
> BTW Re-run from Checks should work.

Hmm, it never seems to work for me. I will try that next time though.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java
Comment (dwnusbaum) [reply src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java:422]: Actually I think the problems were more limited. If you enter this method with code like `MyClass.staticField = x`, then `GroovyCallSiteSelector.field` could only find fields declared in `java.lang.Class` itself or its supertypes, so `field != null` could only be true if your static field has the same name as a field declared in `java.lang.Class`. In other cases, the code will fall through to the new case from https://github.com/jenkinsci/script-security-plugin/commit/85d16bd851b355458f6396dbc4f92c3e0eebcd72 (which has test coverage), and will work correctly. So for example, if you have a static field named `name`, `MyClass.@name = x` will fail because `field java.lang.Class name` is not allowed to be used. It didn't work correctly before https://github.com/jenkinsci/script-security-plugin/commit/85d16bd851b355458f6396dbc4f92c3e0eebcd72 either though, so impact is minimal.

The other difference would be that if you had code like `object.staticField` for some field that you had approved as `staticField Foo foo` previously, the old code would check if `field Foo foo` was approved, and if not, it would reject it as `field ...`, rather than checking if `staticField ...` is allowed. That also would have been broken before https://github.com/jenkinsci/script-security-plugin/commit/85d16bd851b355458f6396dbc4f92c3e0eebcd72 though.

**Pull request merged**
